src/runtime/hexagon/hexagon_buffer.cc

```diff
@@ -49,6 +49,16 @@ struct Allocation {
 struct DDRAllocation : public Allocation {
   DDRAllocation(size_t nbytes, size_t alignment) : Allocation(nbytes, alignment) {
     int ret = posix_memalign(&data_, alignment, nbytes);
+
+    // The heap used by malloc on Hexagon is always mapped as cacheable. The heap manager may
+    // not perform cache flush and invalidation on a prior memory free. So, a subsequent memory
+    // allocation request to the heap manager may allocate memory that resides in part or in full in
+    // the cache. Hence, we must flush and invalidate the allocation from the cache to ensure that
+    // DMA with cache bypass enabled will function properly. DMA with cache bypass enabled assumes
+    // that HexagonBuffer objects are not cached unless explicitly modified by the primfunc. We must
+    // flush and invalidate after malloc to uphold this assumption.
+    qurt_mem_cache_clean(reinterpret_cast<qurt_addr_t>(data_), nbytes,
+                         QURT_MEM_CACHE_FLUSH_INVALIDATE, QURT_MEM_DCACHE);
```

Comment:
Could this be invalidate only, since there is no data copied in this buffer yet?

Reply:
Yes, and, by the same logic we should be able to invalidate (only) the source of the `memcpy` below given that the source is not modified by the `memcpy`.  I have made this change as well.

Reply:
Without going into too much detail, I believe that invalidate (only) maps to "flush and invalidate" when data is modified in the cache.  Before this PR we were doing invalidate (only) on the destination of the `memcpy` which is modified in cache which should not have worked unless the modified data was first flushed from the cache.  Still, using "invalidate" for unmodified data and "flush and invalidate" for modified data is semantically correct and makes the code more readable so I will follow that convention.